Spencer

sjkillen opened the pull request.

r30shah committed:
Add options for compressing AOT Code

This commit adds following options to control and trace compression of
AOT code.
1. disableAOTBytesCompression : Disable compressing AOT compiled method
   before storing it into cache.
2. verbose={compression} : Support verbose out put about compressing/decompressing
   AOT compiled method in the verbose log.

Signed-off-by: Rahil Shah <rahil@ca.ibm.com>

yanluo7 committed:
Add jit options to trace/disable generic static final field fold

Signed-off-by: Yan Luo <Yan_Luo@ca.ibm.com>

Commit:
Use symRef with known obj index for known parms

If known object information is avaible for the method receiver, use
known object index to create the parm symbol reference. Optimizations
can make use of the info without running global VP which is very
expensive and doesn't kick in at warm.

Signed-off-by: Yi Zhang <yizhang@ca.ibm.com>

amicic committed:
Flag to control use of GC startup hints

Introduce an explicit flag to control use of heap sizing startup hints
from Shared Cache. Disabled by default.

Restore internal learning parameter that was set to 0 (temporarily used
as a shortcut to disable the feature).

Signed-off-by: Aleksandar Micic <amicic@ca.ibm.com>

amicic committed:
Remove expensive CS assert

The assert served its purpose. Removing it now, since it's defeating
leaf object optimization.

Signed-off-by: Aleksandar Micic <amicic@ca.ibm.com>

amicic committed:
Concurrent Mark kickoff with Concurrent Scavenger

Normally, we prevent CM kickoff logic before Scavenge Stat info is
available (survival rate, allocate space size....). In presence of CS,
we incorrectly conclude that first Scavenge cycle is complete (while
only the first STW increment is complete) and proceed with incomplete
data to kickoff math. This indirectly may lead to premature and too
frequent kickoff.

This problem is only visible when CM concurrentSlack option is provided
with high values.

Signed-off-by: Aleksandar Micic <amicic@ca.ibm.com>

amicic committed:
Allocation deviation boost in CS

When calculating tilt ratio in Concurrent Scavenger, allocation rate is
taken into account (beside Nursery flipping rate), since during active
CS cycle allocation is fed from common allocate/survivor space.

Allocate rate can vary a lot (more than flipping rate). So far, we
compensated the variability by boosting the average allocate rate by
fixed 20%. Real life variability could be much larger.

We now track deviations in allocation rate from the average, and use
this parameter to more precisely boost the allocation rate provided to
the tilt math.

Signed-off-by: Aleksandar Micic <amicic@ca.ibm.com>

Commit:
Create known object temps when splitting block

When splitting a block, temps are created to uncommon references across
the splitted blocks. Use known object temps, if the original referenced
value is a known object.

Signed-off-by: Yi Zhang <yizhang@ca.ibm.com>

yanluo7 committed:
Tril tests for ternary opcode

Add tril tests for ternary opcode with different compare nodes and value
nodes

Signed-off-by: Yan Luo <Yan_Luo@ca.ibm.com>

knn-k committed:
AArch64: Add some members to ARM64LinkageProperties

This commit adds some members and their getters to ARM64LinkageProperties
for use in OpenJ9.

Signed-off-by: knn-k <konno@jp.ibm.com>

fjeremic committed:
Enable GTest <tuple> on z/OS

The C++11 `tuple` class can be included in GTest for z/OS however it is
part of the TR1 extensions, and hence lives in the TR1 namespace. GTest
is configured as such so it expects `tuple` to be in the `std`
namespace. There is a define override (`GTEST_ENV_HAS_STD_TUPLE_`)
however which can be used to import the `tuple` class on z/OS.

We make use of this override along with a code change to enable the use
of system `tuple` class in GTest and it's consumers (Tril).

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Allow use of _ISOC99_SOURCE and __IBMCPP_TR1__ on the command line

Some existing pieces of code attempt to define this macro which enables
us to use specific APIs on z/OS such as `snprintf`, however they are
not checking if the define already exists on the command line which
results in redefinition errors. Fix this by checking if the define
already exists.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Define _ISOC99_SOURCE for all z/OS compiles and remove IBM_ATOE

No reason not to include this define to expose APIs such as `snprintf`
globally. We also remove the `IBM_ATOE` define since it is specific to
ASCII to EBCDIC conversions of files not checked out by Git for z/OS
which is no longer the case.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Define YYLMAX for Lex token limit and clean up xlc compiler directives

`YYLMAX` is a define which is used to override the default token length
limit on z/OS. Because of the copyright headers the token length of a
comment may exceed the default value (100) and Lex will fail to parse
the comment token. We increase the value to 1000 here to avoid such an
issue.

We also take this opportunity to clean up some xlc compile flags which
are not useful or detrimental to compilation of OMR sources on z/OS.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Add j9a2e as a link target of comptest

The comptest module consumes omrport which relies on the atoe.h APIs
exposed. These APIs are defined in the j9a2e library so we must add it
as a link target.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Change BC to an RX-a instruction format

Technically this is an RX-b instruction however we currently do not
have a proper class for such instructions and we generate the `BC`
instruction as part of z/OS linkage in which we use a real register to
generate the mask. Assertions later on prevent us from doing this if
the `BC` instruction is properly marked as an RX-a instruction.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Relax C++11 usage within Tril tests cases to accommodate z/OS

z/OS cannot handle some of the C++11 features currently used within
Tril such as brace initialization and some flavours of tuple type
reduction.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Guard ASCII to/from EBCDIC runtime conversions with !defined(OMR_EBCDIC)

There are many places in OMR which on the z/OS platform attempt to
perform ASCII/EBCDIC conversions at runtime for various library
calls. These conversions are needed because some consumers of OMR
encode the files on disk in one format but expect it at runtime in
another format.

A lot of this is really not needed had the files been properly tagged
at the filesystem level via `chtag`. However this is not the case
today for the downstream projects.

In OMR however we are not transitioning to using Git on z/OS which
does properly tag all files once checked out of the repository. As
such no runtime ASCII/EBCDIC conversion is needed, and hence all the
paths which do runtime conversions need to be properly guarded since
we export the `OMR_EBCDIC` flag on all compilation units on z/OS.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Export OnLoad* symbols for all agent test libraries

For platforms which require "extra" metadata for exported library
functions (Windows and z/OS) we need to call the helper routine to
generate such metadata so that we are able to load the symbols at
runtime.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Install omrsig import library alongside the shared library

When building omrsig on z/OS the build process fails because the import
library (libomrsig.x) does not live alongside the shared library
(libomrsig.so). Because of this the make process fails as we cannot
find the import library.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Relax C++11 usage in JitBuilder tests for z/OS

Initializer lists of vectors do not work on z/OS due to compiler
restrictions.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Avoid use of std::to_string as z/OS does not support it

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Avoid death tests as ExitedWithCode is not defined on z/OS

Because `GTEST_HAS_DEATH_TEST` is not defined on z/OS neither is the
`ExitedWithCode` API, and hence we cannot make use of it.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Relax usage of C++11 in compilertests and tril/test

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Use CMake build system for z/OS PR builds

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

Commit:
CMake: Refactor DDR support

Introduced a number of changes

- Headers can now be directly associated with a ddr set
- A DDR set no longer needs to be specified when enabling DDR for a target.
- Added a new command `ddr_set_add_targets` which adds a target to a DDR
  set. Combined with the above, this allows for a delayed definition of a
  DDR set. In addition targets can be in multiple DDR sets.
- DDR sets can now be composed via `ddr_set_add_targets`. This means we can
  get rid of the awkward `OMR_DDR_SET` mechanics. Instead we just define
  our own DDR set, and consumers simply add our DDR set to theirs
- omr_ prefix removed from ddr related target names.
- Added GLOB_HEADERS and GLOB_HEADERS_RECURSIVE options to
  target_enable_ddr, which glob the target source directory for headers

Signed-off-by: Devin Nakamura <devinn@ca.ibm.com>

fjeremic committed:
Enable -DGTEST_ENV_HAS_STD_TUPLE_ only on z/OS and fix copyrights

This is causing AIX problems:

```
/third_party/gtest-1.8.0/include/gtest/internal/gtest-port.h", line
714.7: 1540-0130 (S) "::std::get" is not declared.
/third_party/gtest-1.8.0/include/gtest/internal/gtest-port.h", line
715.7: 1540-0130 (S) "::std::make_tuple" is not declared.
/third_party/gtest-1.8.0/include/gtest/internal/gtest-port.h", line
716.7: 1540-0130 (S) "::std::tuple" is not declared.
/third_party/gtest-1.8.0/include/gtest/internal/gtest-port.h", line
717.7: 1540-0130 (S) "::std::tuple_element" is not declared.
/third_party/gtest-1.8.0/include/gtest/internal/gtest-port.h", line
718.7: 1540-0130 (S) "::std::tuple_size" is not declared.
```

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

gacholio committed:
Runtime compressed refs work

Signed-off-by: Graham Chapman <graham_chapman@ca.ibm.com>

fjeremic committed:
Remove UT_DIRECT_TRACE_REGISTRATION define from z/OS

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

fjeremic committed:
Fixup whitespace, comments, and additional copyrights in LEX and YACC

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

babsingh committed:
New OMRPORT_SIG_FLAG_* macros to extend signal support on Unix platforms

The new macros support the unix/omrsignal.c implementation, which is
used on OSX, Linux, AIX and zOS.

The new macros allow us to support 27 asynchronous and synchronous
signals each.

Co-authored-by: Keith W. Campbell <keithc@ca.ibm.com>
Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Fix formatting for OMRPORT_SIG_FLAG_* macros

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
omrsig_can_protect should only support synchronous signals

omrsig_can_protect is used to check if omrsig_protect can be used.
omrsig_protect only works with synchronous signals. omrsig_can_protect
does not need to check for asynchronous signals because omrsig_protect
does not work with asynchronous signals. So, asynchronous signal checks
have been removed from omrsig_can_protect.

The function description of omrsig_can_protect has also been updated.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Fix formatting and consistently use tracepoints for omrsig_can_protect

Tracepoints have been consistently introduced in each platform
implementation of omrsig_can_protect.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Update omrsig_can_protect stub

omrsig_can_protect stub supports no signals so it should never return 1.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Fix ifdef formatting as per the coding standards

In addition to fixing ifdef formatting,
- few unneeded ifdefs were removed.
- comment labels were fixed in few cases.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Split bit-masks to represent sync and async signals separately

The following bit-masks are split into two separate bit-masks:
1. signalsWithHandlers split into syncSignalsWithHandlers and
asyncSignalsWithHandlers.
2. signalsWithMasterHandlers split into syncSignalsWithMasterHandlers
and asyncSignalsWithMasterHandlers.

Synchronous and asynchronous signals require separate bit-masks because
the sync and async signal flags will overlap from now onwards in
unix/omrsignal.c.

The old bit-masks (signalsWithHandlers and signalsWithMasterHandlers)
will be removed once the functions in unix/omrsignal.c are updated.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Add set and unset functions for the newly split bit-masks

Set and unset functions have been added for the newly split
bit-masks: [sync|async]SignalsWithHandlers and
[sync|async]SignalsWithMasterHandlers.

These functions check if a port library signal flag is synchronous or
asynchronous, and then accordingly set/unset the corresponding bit-mask.

Added a macro to mask the control bits in the port library signal flag.

While performing an unset operation, the shared control flags are never
unset. This allows to easily check if a port library signal is set in
the bit-mask. 

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Use the new set and unset functions for the updating new bit-masks

This removes the usage of the old bit-masks, and brings us one step
closer to removing the old bit-masks.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Update usage of OMR_IS_ONLY_ONE_BIT_SET in unix/omrsignal.c

omrsig_set_single_async_signal_handler, omrsig_register_os_handler and
omrsig_is_signal_ignored check if only one port library signal flag is
set in the input.

Now, the port library signal flags in unix/omrsignal.c will always have
more than one bit set. One signal bit and one control bit can be set in
the port library signal flag.

OMR_IS_ONLY_ONE_BIT_SET checks have been updated to verify if only one
non-control bit is set in the port library flag. To accomplish this, the
control bits are masked before invoking OMR_IS_ONLY_ONE_BIT_SET.

The definition of OMR_IS_ONLY_ONE_BIT_SET has been fixed.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Update omrsig_set_options in unix/omrsignal.c

- Use macros such as OMR_ARE_[NO|ANY|*]_BITS_SET.
- Use the new bit-masks, [sync|async]SignalsWithHandlers, instead of the
old one, signalsWithHandlers.
- Fix comments; avoid using first person.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Update registerMasterHandlers in unix/omrsignal.c

- Use the new bit-masks, [sync|async]SignalsWithMasterHandlers, instead
of the old one, signalsWithMasterHandlers.
- Reformat the code and comments to abide the coding standards.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Update OMR_ARE_ANY_BITS_SET checks related to the signal flag

In unix/omrsignal.c, the port library signal flags can have more than
one bit set. So, OMR_ARE_ANY_BITS_SET operations between the port
library signal flag and OMRPORT_SIG_FLAG_SIGALL[SYNC|ASYNC] are no
longer safe since OMR_ARE_ANY_BITS_SET does not verify if all the bits
in the port library signal are set in the synchronous or asynchronous
signal-set bit-mask. Replacing OMR_ARE_ANY_BITS_SET with
OMR_ARE_ALL_BITS_SET and interchanging the order of the parameters in
OMR_ARE_ANY_BITS_SET yields a more safer (correct) check.

Also, the code and comments have been reformatted as per coding
standards.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Check if a set of signal flags is ambiguous

In unix/omrsignal.c,
- Some omrsig* functions can accept multiple signal flags.
- No omrsig* function accepts both asynchronous and synchronous signals
in the same function call. In other words, an omrsig* can accept either
a set of asynchronous or synchronous signals.

If asynchronous and synchronous signal flags are bitwise-or'ed, then
this will yield an ambiguous set of signal flags. In the omrsig*
functions that accept signal flags as input, ambiguity checks have been
added. These checks will return an error if an ambiguous set of signal
flags is identified.

flags=0 is valid in some cases, and it indicates cleanup/removal of a
handler. The ambiguity check does not return an error for flags=0.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Use syncSignalsWithMasterHandlers in omrsig_protect [unix/omrsignal.c]

In unix/omrsignal.c::omrsig_protect, use the new bit mask,
syncSignalsWithMasterHandlers, instead of the old bit-mask,
signalsWithMasterHandlers.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Remove the definition of the old bit-masks in unix/omrsignal.c

The usage of the old bit-masks, signalsWithHandlers and
signalsWithMasterHandlers, has been replaced with the usage of the new
bit-masks, [sync|async]SignalsWithHandlers and
[sync|async]SignalsWithMasterHandlers. Thus, the declaration of the old
bit-masks is removed.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh committed:
Add support for SIGTTIN and SIGTTOU in unix/omrsignal.c

Support for the last two POSIX signals, SIGTTIN and SIGTTOU, has been
added.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

wbh123456 committed:
Remove all references to bucmpeq

A test to see if the code still works by removing only references to
bucmpeq and ifbucmpeq.

Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

babsingh committed:
Update masterSynchSignalHandler to handle all SIGFPE variants

thisRecord->flags will only have OMRPORT_SIG_FLAG_SIGFPE set since the
SIGFPE variants are not included in the OMRPORT_SIG_FLAG_SIGALLSYNC
bit-mask. In masterSynchSignalHandler, the received signal can be a
variant of SIGFPE: DIV_BY_ZERO, INT_DIV_BY_ZERO or INT_OVERFLOW.

masterSynchSignalHandler should handle all the SIGFPE variants if
thisRecord->flags has OMRPORT_SIG_FLAG_SIGFPE set. In
masterSynchSignalHandler, the received signal is converted into the port
libray signal flag form and stored in portLibType. If portLibType is a
SIGFPE variant, then it is converted into OMRPORT_SIG_FLAG_SIGFPE. This
will allow masterSynchSignalHandler to handle all SIGFPE variants if
thisRecord->flags has OMRPORT_SIG_FLAG_SIGFPE set.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

wbh123456 committed:
Remove references of deprecated IL opcode TR:bucmpne and TR:iucmpeq

Remove all uses of the two variables in OMR.

Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

Commit:
Tweak CMake handling on z/OS

- Better identify z/OS specific config rather than xlc + zArch
- Allow ability to specify independently control EBCDIC usage in tools
  and runtime components

Signed-off-by: Devin Nakamura <devinn@ca.ibm.com>

Commit:
Remove unnecessary references to J9ZOS390 in z/OS specific files

Signed-off-by: Devin Nakamura <devinn@ca.ibm.com>

yanluo7 committed:
Enhances CFG simplification with new patterns

- an explicit if obj == null to a throw of new NullPointerException will
  now be simplified to NULLCHK (for both resolved and unresolved
  NullPointerException cases.
- an explicit if obj == null to a throw of another exception will become
  a NULLCHK to a catch with a goto to the throwing code
- a diamond or triangle storing a single temp will become a ternary
- a triangle storing to multiple temps will become a sequence of
  ternaries
- an excplicit if obj instanceOf class with a throw becomes a checkcast
  to a catch with a goto to the throwing code

This commit also adds CFGSimplification to ILGenOpts to fold complex
control flow early

Co-authored-by: Andrew Craik <ajcraik@ca.ibm.com>

Signed-off-by: Yan Luo <Yan_Luo@ca.ibm.com>

andrewcraik committed:
Improve compactNULLCHKs

This change does two things: (1) when we find a NULLCHK with a first
child that has reference count 1 - turn that node into a PassThrough to
provide an opportunity for something more useful to move up under the
NULLCHK and (2) allow compactNULLCKs to cross block boundaries when the
two blocks are in treetop order and are in a single predecessor to
single successor relationship (eg the break is due to exception edges or
similar.

Signed-off-by: Andrew Craik <ajcraik@ca.ibm.com>

wbh123456 committed:
Remove references of deprecated IL opcode TR:iucmpne

removed all uses to iucmpne.

Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

wbh123456 committed:
Remove references of deprecated IL opcode TR:lucmpeq

Remove all uses to lucmpeq in OMR.

Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

babsingh committed:
Replace OMR_ARE_ANY_* with OMR_ARE_ALL_* in unix/omrsignal.c

In unix/omrsignal.c, the signal flags can have more than one bit set.
This makes OMR_ARE_ANY_BITS_SET checks incorrect. All bits need to
be checked to see if a signal flag is set. So, OMR_ARE_ANY_BITS_SET
is replaced with OMR_ARE_ALL_BITS_SET in checks related to the signal
flags.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

wbh123456 committed:
Remove references of deprecated IL opcode TR:lucmpne

removes all uses to lucmpne in OMR.

Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

0xdaryl committed:
Merge pull request #3946 from midronij/j2prof

Remove j2prof

wbh123456 committed:
Remove references of deprecated IL opcode TR:sucmpeq

Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

cedrichansen committed:
Splitting up LargeObjectAllocateStats into two files

- `LargeObjectAllocateStats` .cpp/.hpp contained class declarations for both `MM_LargeObjectAllocateStats` and `MM_FreeEntrySizeClassStats`
- Created new files called `FreeEntrySizeClassStats` (.cpp/.hpp) which contain relevant info for `MM_FreeEntrySizeClassStats`

Signed-off-by: Cedric Hansen <cedric.hansen@ibm.com>

charliegracie committed:
Merge pull request #3955 from cedrichansen/master

Splitting up LargeObjectAllocateStats into two files

Leonardo2718 committed:
Merge pull request #3940 from wbh123456/bucmpeq_test

Remove all uses of the ILOpcode: bucmpeq

wbh123456 committed:
Remove references of deprecated IL opcode TR:bucmpne and TR:ifbucmpne

Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

wbh123456 committed:
Remove references of deprecated IL opcode TR:ifiucmpeq

Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

wbh123456 committed:
Remove references of deprecated IL opcode TR:ifiucmpne

Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

wbh123456 committed:
Remove references of deprecated IL opcode TR:iflucmpeq

Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

wbh123456 committed:
Remove references of deprecated IL opcode TR:iflucmpne

Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

RSalman committed:
Deep Structure Scan Optimization

- deepScan routine introduced to "special treat" objects with self
referencing fields (fast traversal of deep structure nodes)
    - Split into two functions `deepScan` and `deepScanOutline`.
Frequently called checks (see lazy start check below) must be inlined
    - deepScanOutline` scans self referencing field of an object follows
it and repeats. If it can't follow through in one direction, it will
attempt to use the second self referencing field (e.g., the next and
previous pointes of double linked lists)
- Lazy Start condition used for gatekeeping - inhibit the special
treatment routine with relatively high probability to skip over most
false positives (shorter lists), while only marginally delay detection
of very deep structures
    -  Lazy Start check (`(uintptr_t)objectPtr & 0x78) == 0`) for
probability 1/16 to special treat
- `approxScanCacheCount` added to Environment to have a thread local
copy which is required for checking throttling condition, optimization
to avoid overflow of the free list
    - If free list is utilized more than 50% deep scan is terminated

Signed-off-by: Salman Rana <salman.rana@ibm.com>

fjeremic committed:
Remove supportsTrapsInTMRegion from OMR (#3928)

Remove supportsTrapsInTMRegion from OMR

Leonardo2718 committed:
Merge pull request #3922 from wbh123456/getint

Remove getInternalPtrMapBit from OMR

wbh123456 committed:
Remove references of deprecated IL opcode TR:ifsucmpeq

Remove all the uses of ifsucmpeq except the function:
intCompareNarrower(node, s, TR::ifsucmpeq, TR::ifscmpeq, TR::ifbcmpeq);

Also left sucmpeqEvaluator untouched.

Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

charliegracie committed:
Merge pull request #3882 from dnakamura/ddr_refactor

CMake: Refactor DDR support

charliegracie committed:
Merge pull request #3964 from amicic/useGCStartupHints_flag

Flag to control use of GC startup hints

wbh123456 committed:
Remove references of deprecated IL opcode TR:sucmpne and TR:ifsucmpne

Left TR::Node *sucmpneSimplifier(TR::Node * node, TR::Block * block,
TR::Simplifier * s)  and  longCompareNarrower(node, s, TR::ificmpne,
TR::ifsucmpne, TR::ifscmpne, TR::ifbcmpne)  unchanged

Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

babsingh committed:
Share code between old and new signal constants

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

efferifick committed:
Create TR::SelectOpt class

Allows for multiplexing between a default optimizations and a another one
conditional upon a JITCmdLineOption.

Issue: #3956.

Signed-off-by: Erick <eochoa@ualberta.ca>

fjeremic committed:
Fix formatting and properly use HOST_BITS on Z

In addition instead of moving files in the CMake build we copy them
instead.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

charliegracie committed:
Merge pull request #3958 from amicic/CM_kickoff_in_presence_of_CS

Concurrent Mark kickoff with Concurrent Scavenger

Leonardo2718 committed:
Merge pull request #3962 from yanluo7/sfff

Add jit options to trace/disable generic static final field fold

charliegracie committed:
Merge pull request #3959 from amicic/CS_allocDeviationBoost

Allocation deviation boost in CS

fjeremic committed:
Enable OMR to build with OMR_USE_NATIVE_ENCODING=OFF but enable it by default

There are a few changes needed to get OMR to be able to build when specifying `OMR_USE_NATIVE_ENCODING=OFF` however this is not the default OMR configuration. By default this variable is set to `ON` so that we use the native encoding of the platform. Projects consuming OMR are able to use non-native encodings by overriding this variable to `OFF`, however the onus is on the downstream projects which do this to correctly make use of the a2e library provided in OMR.

Signed-off-by: Filip Jeremic <fjeremic@ca.ibm.com>

amicic committed:
Fix loss of precision warning in alloc boost logic

Fix Windows compiler warnings in
https://github.com/eclipse/omr/pull/3959

Signed-off-by: Aleksandar Micic <amicic@ca.ibm.com>

0xdaryl committed:
Merge pull request #3902 from midronij/entry-bps

Generate entry breakpoints for methods compiled with system linkage

0xdaryl committed:
Merge pull request #3898 from wbh123456/dummytemp

Remove _dummyTempStorageRefNode from OMR

charliegracie committed:
Merge pull request #3980 from amicic/CS_allocDeviationBoost

Fix loss of precision warning in alloc boost logic

charliegracie committed:
Merge pull request #3976 from amicic/remove_expensive_CS_assert

Remove expensive CS assert

charliegracie committed:
Merge pull request #3978 from gacholio/compress

Runtime compressed refs work

rwy7 committed:
Turn on hybrid arraylets by default

In the near future, hybrid arraylets (and arraylets in general) will be
permanently enabled. As a start, turn on hybrid arraylets by default.
Arraylets are already enabled.

Signed-off-by: Robert Young <rwy0717@gmail.com>

bragaigor committed:
Fix broken links in OMR README

Signed-off-by: Igor Braga <higorb1@gmail.com>

sjkillen committed:
Add IDT

Leonardo2718 committed:
Merge pull request #3982 from bragaigor/broken_links

Fix broken links in OMR README

charliegracie committed:
Merge pull request #3981 from rwy0717/arraylets

Turn on hybrid arraylets by default

rwy7 committed:
Merge the arraylets and hybrid arraylets build flags

There's only one instance of OMR_GC_HYBRID_ARRAYLETS in the codebase.
Remove the ifdef, so the flag is permanently in effect.

Signed-off-by: Robert Young <rwy0717@gmail.com>

keithc-ca committed:
Use the C++ compiler to preprocess C++ code for DDR

This avoids a problem on z/OS where c89 refuses to consider C++ code:
  FSUM3008 Specify a file with the correct suffix (.c, .i, .s, .o, .x,
.p, .I, or .a), or a corresponding data set name, instead of
RegionPool.cpp.

Signed-off-by: Keith W. Campbell <keithc@ca.ibm.com>

aarongraham9 committed:
GCMaps and RegisterMaps for the AArch64 Compiler

GCMaps and RegisterMaps for the AArch64 Compiler
-Iterate Over IL Instructions and Set a GC Map if Applicable
-Implement RegisterMaps for the AArch64 Compiler

Signed-off-by: Aaron Graham <aaron.graham@unb.ca>

Leonardo2718 committed:
Merge pull request #3693 from fjeremic/jitbuilder-zos

Migrate z/OS to a CMake build system

Commit:
CMake: blobreader needs to link against the thread library

Signed-off-by: Devin Nakamura <devinn@ca.ibm.com>

0xdaryl committed:
Merge pull request #3828 from CAS-Atlantic/aarch64_gc_maps

Implement AArch64  GCMaps

knn-k committed:
AArch64: Fix comment for OMRMemoryReference

This commit fixes the doxygen comment for
OMRMemoryReference::generateBinaryEncoding() for AArch64.

Signed-off-by: knn-k <konno@jp.ibm.com>

nidhi1605 committed:
Remove redundant/unneeded stack map and atlas methods

Fixes #2619

Removed redundant / unneeded stack map and atlas  methods from JitBuilder's FrontEnd and fvtest

Signed-off-by: Nidhi Gupta <itsnidhi16@gmail.com>

0xdaryl committed:
Merge pull request #3985 from knn-k/aarch64memref8

AArch64: Fix comment for OMRMemoryReference

rwy7 committed:
Permanently enable the arraylets flag in OMR

- Add deprecation notes to the configure options
- Permanently enable the ifdefs

Signed-off-by: Robert Young <rwy0717@gmail.com>

vijaysun-omr committed:
Merge pull request #3947 from yanluo7/cfg

Enhances CFG simplification with new patterns

Commit:
Return if MethodHandle.invokeExact has been specialized

Archetype specimen is the most refined form of `MethodHandle.invokeExact`,
thus it does not need to go through any more refinement. Plus its symbol
reference does not have a valid vtable offset, it will cause the
subsequent code that uses this offset to get a resolved method to crash.

Signed-off-by: liqunl <liqunl@ca.ibm.com>

Commit:
CMake: Add ability to skip parsing debug info for ddr

Add NO_DEBUG_INFO option which stops ddrgen from extracting info from
debug symbols for the target

Signed-off-by: Devin Nakamura <devinn@ca.ibm.com>

Commit:
CMake: add headers to scan for macro information

Signed-off-by: Devin Nakamura <devinn@ca.ibm.com>

0xdaryl committed:
Merge pull request #3970 from wbh123456/eqcompare

Remove all uses of code in Equality compare/Equality compare and branch categories

youngar committed:
Add concurrent copy forward read barrier type

Signed-off-by: Andrew Young <youngar17@gmail.com>

youngar committed:
Add option to enable concurrent copy forward

Signed-off-by: Andrew Young <youngar17@gmail.com>

agarwalrounak committed:
Remove GlobalLabel symbols from compiler

Fixes #3993

Signed-off-by: Rounak Agarwal <rounakr.ag73@gmail.com>

nidhi1605 committed:
Remove J9_VERSION and update copyright header

Removed J9_Version from Test compiler and JitBuilder and updated the copyright header to current year.

Fixes #1651

Signed-off-by: Nidhi Gupta <itsnidhi16@gmail.com>

nidhi1605 committed:
Removed RegisterITF references and commented out available bits

Fixes #3961

Signed-off-by: Nidhi Gupta <itsnidhi16@gmail.com>

charliegracie committed:
Merge pull request #4013 from keithc-ca/ddr

Use the C++ compiler to preprocess C++ code for DDR

Commit:
Skip estimate code size for force inlining targets

Limits set by exceedsSizeThreshold are ignored for force inlining
targets. The bytecode size for those targets won't be used so
skip calling estimateAndRefineBytecodeSize for them to save some
compilation time.

Signed-off-by: Yi Zhang <yizhang@ca.ibm.com>

charliegracie committed:
Merge pull request #3988 from rwy0717/arraylets

Merge the arraylets and hybrid arraylets build flags

charliegracie committed:
Add Robert as a code owner to the appropriate files/folders

Signed-off-by: Charlie Gracie <charlie.gracie@gmail.com>

mstoodle committed:
Merge pull request #4014 from charliegracie/codeowners

Add Robert as a code owner to the appropriate files/folders

rwy7 committed:
Merge pull request #3991 from dnakamura/ddr_nodbg

CMake: Add ability to skip parsing debug info for ddr

rwy7 committed:
Merge pull request #3984 from dnakamura/blobreader

CMake: blobreader needs to link against the thread library

nidhi1605 committed:
Remove FlattenRegisterPairs

Fixes #3968

Signed-off-by: Nidhi Gupta <itsnidhi16@gmail.com>

knn-k committed:
AArch64: Implement ARM64HelperCallSnippet

This commit implements ARM64HelperCallSnippet for AArch64.

Signed-off-by: knn-k <konno@jp.ibm.com>

jjohnst-us committed:
Updates associated with z/TPF GCC 7 Builds

Compiler Option updates as part of building for GCC 7.  Additionally,
needed to bypass z/TPF stack validation.

Signed-off-by: James D Johnston <jjohnst@us.ibm.com>

Commit:
Disable boolean store simplification

Boolean store simplification has caused test failures in downstream
project openj9. We haven't fully understood the defect, so disable this
opt until we have a fix for it.

Signed-off-by: liqunl <liqunl@ca.ibm.com>

0xdaryl committed:
Merge pull request #4017 from liqunl/cfgsimplifier

Disable boolean store simplification

keithc-ca committed:
General z/OS improvements

* heed compiler advice about built-in functions:
	INFORMATIONAL CCN3573 ./zos390/omrgetjobname.c:76
	To use the built-in form of the strcpy function add the
	#include <string.h> directive.

Signed-off-by: Keith W. Campbell <keithc@ca.ibm.com>

Leonardo2718 committed:
Merge pull request #3785 from yanluo7/ternarytest

Tril tests for ternary opcode

keithc-ca committed:
Extend DDR rules to handle *.mc files

Fixes error on z/OS:
  No rule to make target 'omrlpdat.i', needed by 'ddrgen'.

Signed-off-by: Keith W. Campbell <keithc@ca.ibm.com>

0xdaryl committed:
Remove inlineNDmemcpyWithPad and fold code assuming false

This function always returns a constant false in OMR.  It is not used
in any known downstream projects.  Remove and fold surrounding code
assuming it returns false.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

0xdaryl committed:
Fold code assuming MemToMemVarLenMacroOp::needsLoop is always true

* Remove the function MemToMemVarLenMacroOp::needsLoop()
* Fold code that references it assuming it is always true

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

youngar committed:
Merge pull request #3992 from dnakamura/ddr_headers2

CMake: add headers to scan for macro information

youngar committed:
Merge pull request #4019 from keithc-ca/zos

General z/OS improvements

youngar committed:
Merge pull request #4020 from keithc-ca/ddr

Extend DDR rules to handle *.mc files

efferifick committed:
Add internal memory to CFG

This commit adds an internal memory region to the CFG.
By default, the internal memory region is created from the heap
region. CFG edges are now allocated to this internal memory
region.

Signed-off-by: Erick Ochoa <eochoa@ualberta.ca>

efferifick committed:
Add optional parameter for TR::CFG

Signed-off-by: Erick Ochoa <eochoa@ualberta.ca>

0xdaryl committed:
Merge pull request #4003 from agarwalrounak/removeGlobalLabel

Remove GlobalLabel symbols from compiler

fjeremic committed:
Misc. Z CodeGenerator cleanups (#4021)

Misc. Z CodeGenerator cleanups

0xdaryl committed:
Remove obsolete code from FEBase.cpp

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

0xdaryl committed:
Remove unnecessary J9_PROJECT_SPECIFIC macros

Code is not required, even for OpenJ9.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

charliegracie committed:
Merge pull request #4016 from jjohnst-us/1902zTPFUpdates

Updates associated with z/TPF GCC 7 Builds

charliegracie committed:
Merge pull request #3995 from youngar/openj9

Add concurrent copy forward read barrier type

wbh123456 committed:
remove unsigned  ILopcode:TR::iflucmpeq and TR::ifiucmpeq from SwitchAnalyzer

Remove the two unsigned opcodes from the SwitchAnalyzer.cpp.

Issue: #3983
Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

charliegracie committed:
Merge pull request #3996 from youngar/new_option

Add option to enable concurrent copy forward

efferifick committed:
Add additional constructor to CFG with region reference parameter.

Eliminate the constructor that takes a region pointer and add a
constructor that takes a region reference as a parameter.
Add optional `TR_AllocationKind` parameter to `addEdge` and
`addExceptionEdge` to allow OpenJ9's master branch to compile.

Signed-off-by: Erick Ochoa <eochoa@ualberta.ca>

sjkillen committed:
Remove IDT from TR_Memory

efferifick committed:
Deprecate the methods `addEdge` and `addExceptionEdge`.

In order to add edges in the CFG to a specific region,
the region must now be specified when constructing the
CFG.

Signed-off-by: Erick Ochoa <eochoa@ualberta.ca>

sjkillen committed:
Add IDT

sjkillen committed:
Neaten up IDT logging

youngar committed:
Merge pull request #3994 from rwy0717/permanently-enable-arraylets

Permanently enable the arraylets flag in OMR

0xdaryl committed:
Rename global register allocated indexes on ParameterSymbol

Provide more meaningful names for the following APIs to make
the code that uses it more readable:

* `getAllocatedIndex` --> `getAssignedGlobalRegisterIndex`
* `setAllocatedIndex` --> `setAssignedGlobalRegisterIndex`
* `getAllocatedLow` --> `getAssignedLowGlobalRegisterIndex`
* `setAllocatedLow` --> `setAssignedLowGlobalRegisterIndex`
* `getAllocatedHigh` --> `getAssignedHighGlobalRegisterIndex`
* `setAllocatedHigh` --> `setAssignedHighGlobalRegisterIndex`

Add Doxygen comments where appropriate.  Modify references in OMR.

The original API is left in place for now while downstream projects
switch over to the new API.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

sjkillen committed:
Catch recursion in IDT

cedrichansen committed:
Modifying kickoff point for concurrent mark

- Added a few heuristics related to the rate of promotion from nursery to better predict how many scavenges remaining until percolatewill occur. These heuristics include a calculation for `headroom` which is related to `concurrentKickoffTenuringHeadroom` and `lastGlobalGCFreeBytes`. Default value for `concurrentKickoffTenuringHeadroom` is 2%
- Kickoff point for concurrent mark will still  be triggered when `scavengesRemaining` point is reached, but this now includes the headroom calculations. The headroom is always set to be 'worst case scenario' between 1, and whatever value is calculated as the headroom remaining
- Added `lastGlobalGCFreeBytesLOA` to `GCExtensionsBase`. The usage of this field is almost identical as `lastGlobalGCFreeBytes`, but only considers LOA. This field is being used in the calculation of `headroom`
- Will be accompanied by changes one the j9 side for parsing `concurrentKickoffTenuringHeadroom` as command line argument

Signed-off-by: Cedric Hansen <cedric.hansen@ibm.com>

klangman committed:
Fix logic in allowGlobalRegisterAcrossBranch() for POWER

The logic in allowGlobalRegisterAcrossBranch() must match the logic in
switchDispatch() when deciding which lookup scheme logic to use. The
failure to check for an overflow in allowGlobalRegisterAcrossBranch()
allowed for GRA to use 25 registers which caused a crash in
freeBestRegister() because all registers were blocked when local RA was
looking to allocate a register needed by the switch lookup code.

Signed-off-by: Kevin Langman <langman@ca.ibm.com>

sjkillen commented:
@andrewcraik

andrewcraik commented on compiler/optimizer/IDT.cpp:
```diff
@@ -0,0 +1,199 @@
+#include "IDT.hpp"
+
+#define INITIAL_NUM_children (10)
```
This is dead I think

andrewcraik commented on compiler/optimizer/IDT.cpp:
```diff
@@ -0,0 +1,199 @@
+void
+IDT::setOnlyChild(IDTNode* owner, IDTNode* child)
+  {
+  owner->_children = (IDTNode::Children*)(void*)((size_t)child | 1);
```
the 1 should be pulled out as a define to give it semantic meaning:
`#define SINGLE_CHILD_BIT 1`
and then
`(IDTNode::Children*)(void*)((size_t)child | SINGLE_CHILD_BIT)`
here you want to cast child to uintptr_t not size_t since size_t is not going to be equal to a pointer width datatype. You should then be able to directly cast to IDTNod::Children

andrewcraik commented on compiler/optimizer/IDT.cpp:
```diff
@@ -0,0 +1,199 @@
+  {
+  if (((size_t)owner->_children) & 1)
+    {
+    return (IDTNode *)(void*)(((size_t)(owner->_children)>>1)<<1);
```
you should be able cast to uintptr_t and bit and with ~SINGLE_CHILD_BIT (see below)

sjkillen replied:
Thanks, I didn't know uintptr_t existed.

andrewcraik commented on compiler/optimizer/IDT.cpp:
```diff
@@ -0,0 +1,199 @@
+IDTNode*
+IDT::getOnlyChild(IDTNode* owner)
+  {
+  if (((size_t)owner->_children) & 1)
```
Use the macro here too (see setOnlyChild)

andrewcraik commented on compiler/optimizer/IDT.hpp:
```diff
@@ -0,0 +1,73 @@
+  next_idx();
+  // Returns NULL if 0 or > 1 children
+  static IDTNode *
+  getOnlyChild(IDTNode *owner);
```
Why aren't these functions on IDTNode?

sjkillen replied:
~~I needed to call them from IDT node for one of the methods. This was my workaround for not being able to call private methods from a friend class~~
Disregard this comment, I posted a follow up to Slack

andrewcraik commented on compiler/optimizer/IDT.hpp:
```diff
@@ -0,0 +1,73 @@
+class IDTNode;
+
+class IDT {
+  friend class IDTNode;
```
If the helpers were on the IDTNode then you wouldn't need this I don't think

sjkillen committed:
Tidy bitmasking _children in IDT

sjkillen committed:
Refactor IDT children accessors

sjkillen committed:
fixup! Refactor IDT children accessors

knn-k committed:
AArch64: Add TR_ARM64countingRecompileMethod to Helpers.inc

This commit adds the symbol TR_ARM64countingRecompileMethod to
Helpers.inc.

Signed-off-by: knn-k <konno@jp.ibm.com>

0xdaryl committed:
Remove CodeCacheBase class

Move the CodeCacheBase fields and functions into the OMR::CodeCache
class, and remove the CodeCacheBase class.  The downstream dependency
requiring this has been eliminated.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

0xdaryl committed:
Merge pull request #4031 from knn-k/aarch64runtime6

AArch64: Add TR_ARM64countingRecompileMethod to Helpers.inc

fjeremic committed:
Rename global register allocated indexes on ParameterSymbol (#4025)

Rename global register allocated indexes on ParameterSymbol

Commit:
Create TR::CFGSimplifier instead of CFGSimplifier

`new CFGSimplifier` in OMR namespace will create an
`OMR::CFGSimplifier` object, so implementation from downstream project will
never be picked up. Fix this by using `TR::CFGSimplifier`.

Signed-off-by: liqunl <liqunl@ca.ibm.com>

0xdaryl committed:
Remove deprecated ParameterSymbol allocated register APIs

* `getAllocatedIndex`
* `setAllocatedIndex`
* `getAllocatedLow`
* `setAllocatedLow`
* `getAllocatedHigh`
* `setAllocatedHigh`

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

cedrichansen committed:
Disable dark matter compact for concurrent sweep

- Disabled dark matter compaction when concurrent sweep is enabled, due to concurrent sweep reporting innacurate dark matter info
- Doubled default value for `darkMatterCompactThreshold` so that it doesn't trigger unless specified by command line option.
- Refer to https://github.com/eclipse/omr/pull/3939 for description of the dark matter compact metrics

Signed-off-by: Cedric Hansen <cedric.hansen@ibm.com>

youngar committed:
Merge pull request #3967 from RSalman/deep-scan

Deep Structure Scan Optimization

0xdaryl committed:
Merge pull request #4035 from liqunl/cfgsimplifier

Create TR::CFGSimplifier instead of CFGSimplifier

0xdaryl committed:
Merge pull request #4027 from cathyzhyi/splitblock

Create known object temps when splitting block

0xdaryl committed:
Merge pull request #3987 from nidhi1605/removeAliasMethod

Remove redundant/unneeded stack map and atlas methods

0xdaryl committed:
Merge pull request #3998 from nidhi1605/removeFlattenRegisterPairs

Remove FlattenRegisterPairs

0xdaryl committed:
Merge pull request #3960 from cathyzhyi/parmsymbol

Use symRef with known obj index for known parms

pdbain-ibm committed:
Handle Windows version 5.1 in omrsysinfo_get_OS_type()

Major verson 5, minor version 1 is Windows XP, 32 bit. An application may
report this version depending on how it is manifested.

Fixes https://github.ibm.com/runtimes/infrastructure/issues/1995

Signed-off-by: Peter Bain <peter_bain@ca.ibm.com>

0xdaryl committed:
Remove deprecated element privatization from FieldPrivatizer

This is a legacy optimization that has never been enabled and not
immediately useful to OMR nor any known downstream projects.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

0xdaryl committed:
Merge pull request #4038 from cathyzhyi/inliner

Skip estimate code size for force inlining targets

0xdaryl committed:
Remove createCoDependentTemporary from SymRefTab

No longer needed now that element privatization has been removed.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

0xdaryl committed:
Remove some unnecessary includes of CS2 files

CS2 files should not be included if they are not required.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

0xdaryl committed:
Merge pull request #4007 from nidhi1605/removeReference

Remove RegisterITF references from OMROptions

fjeremic committed:
Merge pull request #4034 from r30shah/compressAOTCode

Add options for compressing AOT Code

youngar committed:
Merge pull request #3921 from babsingh/omrsig_unix_final_v1

Extend support to all POSIX signals on Unix platforms

0xdaryl committed:
Remove some unnecessary includes of CS2 files

CS2 files should not be included if they are not required.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

0xdaryl committed:
Remove NonUserMethod enum

The enum has been relocated to OpenJ9.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

0xdaryl committed:
Remove unnecessary includes and forward decls from OMRMethod.hpp

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

0xdaryl committed:
Merge pull request #4015 from knn-k/aarch64codegen4

AArch64: Implement ARM64HelperCallSnippet

0xdaryl committed:
Merge pull request #3945 from andrewcraik/compactNULLCHKs

Improve compactNULLCHKs

fjeremic committed:
Remove some unnecessary includes of CS2 files (#4041)

Remove some unnecessary includes of CS2 files

vijaysun-omr committed:
Merge pull request #4024 from 0xdaryl/feclean

Misc. compiler code cleanup

fjeremic committed:
Remove some unnecessary includes of CS2 files (#4046)

Remove some unnecessary includes of CS2 files

0xdaryl committed:
Merge pull request #4026 from klangman/PR141828

Fix logic in allowGlobalRegisterAcrossBranch() for POWER

Leonardo2718 committed:
Merge pull request #4036 from 0xdaryl/linkagecleanup

Remove deprecated ParameterSymbol allocated register APIs

sjkillen committed:
refactor IDT

andrewcraik commented on compiler/optimizer/IDT.hpp:
```diff
@@ -1,73 +1,56 @@
+class IDT
+  {
+  private:
+  class Node
```
Will hiding the node inside cause any problems for Erick's follow-on work?

0xdaryl committed:
Create TR::Method typedef and modify TR_Method references

Prepare to make the `TR_Method` base class an extensible class.  Introduce
a typedef of `TR_Method` to `TR::Method` and modify references throughout
the OMR compiler.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

fjeremic committed:
Misc. OMRMethod.hpp cleanup (#4045)

Misc. OMRMethod.hpp cleanup

youngar committed:
Merge pull request #4051 from pdbain-ibm/version

Handle Windows version 5.1 in omrsysinfo_get_OS_type()

wbh123456 committed:
Remove deprecated const IL Opcodes from SwitchAnalyzer

Remove deprecated const unsigned IL Opcodes from SwtichAnalyzer.cpp.
Removed: TR::iuconst and TR::luconst

Issue: #3983
Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

vijaysun-omr committed:
Merge pull request #4023 from wbh123456/switch

Remove unsigned ILopcode: TR::iflucmpeq and TR::ifiucmpeq from SwitchAnalyzer

knn-k committed:
Arch64: Enable compressedRefsEvaluator

This commit adds compressedRefsEvaluator to the TreeEvaluatorTable
for AArch64.  It calls the platform-independent
OMR::TreeEvaluator::compressedRefsEvaluator().

Signed-off-by: knn-k <konno@jp.ibm.com>

0xdaryl committed:
Add Andrew Craik to CODEOWNERS file

Co-owner of optimizer directory.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

Leonardo2718 committed:
Merge pull request #4058 from 0xdaryl/codeowner

Add Andrew Craik to CODEOWNERS file

Leonardo2718 committed:
Merge pull request #4054 from 0xdaryl/trmethod

Create TR::Method typedef and modify TR_Method references

vijaysun-omr committed:
Merge pull request #4040 from 0xdaryl/noelementprivatization

Remove deprecated element privatization from FieldPrivatizer

knn-k committed:
AArch64: Enable beginInstructionSelection() for PrivateLinkage

This commit removes the line "TR_UNIMPLEMENTED();" for TR_Private
from OMR::ARM64::CodeGenerator::beginInstructionSelection().

Signed-off-by: knn-k <konno@jp.ibm.com>

Leonardo2718 committed:
Merge pull request #4005 from nidhi1605/removej9

Remove J9_VERSION from JitBuilder and Test Compiler

0xdaryl committed:
Merge pull request #4066 from knn-k/aarch64codegen5

AArch64: Enable beginInstructionSelection() for PrivateLinkage

youngar committed:
Merge pull request #4050 from cedrichansen/concurrentKickoff

Modifying kickoff point for concurrent mark

0xdaryl committed:
Merge pull request #4057 from knn-k/aarch64eval41

Arch64: Enable compressedRefsEvaluator

0xdaryl committed:
Merge pull request #3787 from knn-k/aarch64linkage6

AArch64: Add some members to ARM64LinkageProperties

charliegracie committed:
Merge pull request #4037 from cedrichansen/darkMatterFix

Disable dark matter compact for concurrent sweep

rwy7 committed:
Handle args in FindLEX/FindYACC

Signed-off-by: Robert Young <rwy0717@gmail.com>

0xdaryl committed:
Remove cacheStringAppend from VPHandlersCommon

Optimization is vestigal and not used.  Remove it and all associated
code and data structures.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

0xdaryl committed:
Remove vestigal rematerializeSSAddress

Left over from a legacy project and not useful to OMR.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

vijaysun-omr committed:
Merge pull request #4049 from liqunl/localopts

Return if MethodHandle.invokeExact has been specialized

vijaysun-omr committed:
Merge pull request #4052 from wbh123456/const_in_switch

Remove deprecated const IL Opcodes from SwitchAnalyzer

vijaysun-omr committed:
Merge pull request #4073 from 0xdaryl/stringappend

Remove cacheStringAppend from VPHandlersCommon

fjeremic committed:
Handle args in FindLEX/FindYACC (#4071)

Handle args in FindLEX/FindYACC

sjkillen committed:
Add doxygen to IDT.hpp

fjeremic committed:
Remove vestigal rematerializeSSAddress (#4075)

Remove vestigal rematerializeSSAddress

sjkillen committed:
Base proposals added

sjkillen committed:
fixup! Base proposals added

fjeremic committed:
Remove CodeCacheBase class (#4032)

Remove CodeCacheBase class

efferifick committed:
Merging

efferifick committed:
Merge branch 'selectOpt' into ericks-main

efferifick committed:
Removing SelectInliner

efferifick committed:
Merge branch 'erick-removeOptionalMemoryKind' into ericks-main

sjkillen committed:
Merge branch 'ericks-main' of github.com:themaplelab/openj9-omr into spencer

efferifick committed:
Removes OMR::ByteCodeCFG and CFGgenerator

Signed-off-by: Erick Ochoa <eochoa@ualberta.ca>

sjkillen committed:
Merge branch 'ericks-main' of github.com:themaplelab/openj9-omr into spencer

efferifick committed:
Merge branch 'spencer' into ericks-main

sjkillen committed:
Use benefit trace flag

efferifick committed:
Merge branch 'spencer' into ericks-main

efferifick committed:
Keep CFG in a region

Signed-off-by: Erick Ochoa <eochoa@ualberta.ca>

efferifick committed:
fixing frequency

efferifick committed:
Merge branch 'Growable_2d_array' into ericks-main

efferifick committed:
Add Growable_2d_array

efferifick committed:
Merge branch 'Growable_2d_array' into ericks-main

efferifick committed:
Removing double initialization

efferifick committed:
Rudimentary Call Ratio

efferifick committed:
Changes needed for running inliner packing

efferifick committed:
before abstract values

efferifick committed:
Adds initial implementation of Abstract Interpreter

sjkillen committed:
Change header

sjkillen committed:
Merge branch 'ericks-main' of github.com:themaplelab/openj9-omr into spencer-temp

efferifick committed:
CFGNodes debug

sjkillen committed:
Merge branch 'ericks-main' of github.com:themaplelab/openj9-omr into spencer-temp

efferifick committed:
abstract semantics

efferifick committed:
fixes memory issues

efferifick committed:
invokes slightly working

efferifick committed:
invokes

sjkillen committed:
Add loop analyzer

sjkillen committed:
Merge branch 'ericks-main' of github.com:themaplelab/openj9-omr into spencer-temp

efferifick committed:
one iteration

sjkillen committed:
Merge branch 'ericks-main' of github.com:themaplelab/openj9-omr into spencer-temp

sjkillen committed:
partial loops

andrewcraik commented on compiler/il/OMRBlock.hpp:
```diff
@@ -478,6 +483,8 @@ class OMR_EXTENSIBLE Block : public TR::CFGNode
     * Flag functions end
     */
 
+   AbsEnvStatic *_absEnv;
```
we don't want to have this pointer in every block of every compile - that is going to be too expensive - we need to find another way to get hold of it.

efferifick replied:
Sounds good! We can talk about this in a couple of minutes.

andrewcraik commented:
The absenv for OMR will need to be abstracted form Java with a suitable API so that an OpenJ9 abstract interpreter can be created to process the bytecodes. This is necessary because OMR is language independent and does not accept language specific implementation.
